[REFACTOR] Split up ctffind

jojoelfe opened the pull request:
# Description

This PR does not change any code. It just splits up ctffind.cpp into several smaller compilation units. The main motivation is to make future PRs that modify the tilt estimation feature and add other features easier to read and make it easier to reason about impacts of these changes.

The only more substantial change is to rename `PixelSizeForFitting()` to `DilatePowerspectrumToNewPixelSize()`, since there was a comment that suggested that this function should be renamed. 

@arohou this will probably also need adjustments to the CMake system. If you are generally on board with this I can maybe look into adding some of the changes from #411 to this.

Fixes # (issue)

# I have rebased my feature branch to be current with the master branch using to minimize conflicts and headaches

- [x] yes
- [ ] no

# Which compilers were tested

- [ ] g++
- [x] icpc
- [ ] clang
- [ ] other (please specify)

# These changes are isolated to the

- [ ] gui
- [ ] core library
- [ ] gpu core library
- [x] program it modifies

# How has the functionality been tested?

Please describe the tests that you ran to verify your changes. Please also note any relevant details for your test configuration.

- [x] Tested manually from GUI
- [ ] Tested manually from CLI
- [ ] Passed console tests
- [ ] Passed samples functional testing
- [ ] other (please specify)

# Checklist:

- [ ] I have not changed anything that did not need to be changed
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, (***w.r.t. why***), particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the [documentation](https://github.com/bHimes/cisTEM_docs) {Ok to pass for now}
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

bHimes commented:
This was also on my radar, both for speeding up tilted estimation, and adding GPU options. I'll take a look tomorrow a.m. as this refactor would (imo) be good to try to do just once. Glad you got it going!

bHimes commented:
Okay, looking through the changes suggested by @jojoelfe, I would categorize them as partially based on separating out functionality. Below are my notes on the current implementation and suggestions:


### CtffindApp (ctffind.cpp)

In addition to its primary function ctffind.cpp also defines:

[Primary Function](#primary) 
1. [CTFTilt class](#ctftilt) 
2. [NikoTestApp](#nikotest)
3. [ImageCTFComparison class](#imgcompare)
4. [CurveCTFComparison class](#curvecompare)
5. [CtffindObjectiveFunction](#imgobj)
6. [CtffindCurveObjectiveFunction](#curvobj)
7. [Image pre-processing methods](#preprocessing)

### Currently

The primary purpose is as a program to take input from the user, find defocus (and perhaps other parameters like iciness and thickness, etc.) and return those fits.

This houses everything needed to return these fits, including code that may be useful elsewhere in cisTEM. For example, during image simulation in wave_function_progator, I needed some methods (since moved to the Image class) to pre-process data to analyze Thon rings.

#### Primary Function of ctffind <a name="primary"></a>

- The first thing needed is an amplitude spectrum. This can be provided by the user (but not if fitting tilt) or it is calculated from an image or movies. The "regular" approach might be considered a specialization of the tilt approach, where we assert the tiltaxis/angle is = 0.0 and as such the composite spectrum has only one tile with zero stretch.
- Note: beyond this point, the only difference with a tilted image is to output the tilt angle. 
- The average power spectrum is then filtered in 2d to improve the downstream fitting
- The main algo then carries out the fit with many options and in several steps to ensure robustness
- The fitting algo depends on externally defined optimizers that depend on internally defined objective functions.
- Output images and text files are created and placed in accessible locations

#### Suggested refactor:

* Retain the main algo that handles all user-specified logic and bounds on what type of fitting to do.
* Retain the definitions for ctf specific objective functions

## CTFTilt <a name="ctftilt"></a>

This class has main functions.

1. determine information about the sample tilt-axis and tilt angle
2. use axis/angle information to create a stretched avg power spectrum to improve the fitting of defocus parameters.

*Note:* Both of these fits require an initial estimate of the defocus/astigmatism. Whereas the main algo uses the mirroring trick to first estimate astigmatism then estimate defocus, the CTFTilt method first estimates defocus and then does a brute force search on astigmatism. There is no information noted about this decision, but I expect it is that the foreshortening of the PS perpendicular to the tilt-axis might make the mirroring trick fail.

#### Suggested refactor:

* Should move to its own compilation unit  (As done by @jojoelfe in this pr intially)
* and be renamed to reflect it's purpose, perhaps TiltedPSNormalizer 

## NikoTestApp <a name="nikotest"></a>

Clearly, this is copypasta and should just be deleted, right?

## ImageCTFComparison class <a name="imgcompare"></a>

This is a common construct in cisTEM and collects and manages objects that need to be passed to an optimizer.

#### Suggested refactor:

* [See here](#optrefactor)

## CurveCTFComparison class <a name="curvecompare"></a>

This is a common construct in cisTEM and collects and manages objects that need to be passed to an optimizer.

#### Suggested refactor:

* [See here](#optrefactor)

## CtffindObjectiveFunction <a name="imgobj"></a>

The thing the comparison object is passed to and in turn, passed to the optimizer.

#### Suggested refactor: <a name="optrefactor"></a>

The objective function used to score some value and the optimizer used to search that space is generally separable from the main efforts of a cisTEM program. They should be replaceable and modifiable outside the guts and possible initial brute force searches.

I think these should be separate classes from ctffind, though perhaps the ObjectiveFunction could be a method of the Comparator class, which would be a bit cleaner IMO.

## CtffindCurveObjectiveFunction <a name="curveobj"></a>

#### Suggested refactor:

* [See here](#optrefactor)

## Image pre-processing methods <a name="preprocessing"></a>

```cpp
float FindRotationalAlignmentBetweenTwoStacksOfImages(Image* self, Image* other_image, int number_of_images, float search_half_range, float search_step_size, float minimum_radius, float maximum_radius);
void  ComputeImagesWithNumberOfExtremaAndCTFValues(CTF* ctf, Image* number_of_extrema, Image* ctf_values);
int   ReturnSpectrumBinNumber(int number_of_bins, float number_of_extrema_profile[], Image* number_of_extrema, long address, Image* ctf_values, float ctf_values_profile[]);
void  ComputeRotationalAverageOfPowerSpectrum(Image* spectrum, CTF* ctf, Image* number_of_extrema, Image* ctf_values, int number_of_bins, double spatial_frequency[], double average[], double average_fit[], double average_renormalized[], float number_of_extrema_profile[], float ctf_values_profile[]);
void  ComputeEquiPhaseAverageOfPowerSpectrum(Image* spectrum, CTF* ctf, Curve* epa_pre_max, Curve* epa_post_max);
void  OverlayCTF(Image* spectrum, CTF* ctf, Image* number_of_extrema, Image* ctf_values, int number_of_bins_in_1d_spectra, double spatial_frequency[], double rotational_average_astig[], float number_of_extrema_profile[], float ctf_values_profile[], Curve* equiphase_average_pre_max, Curve* equiphase_average_post_max);
void  ComputeFRCBetween1DSpectrumAndFit(int number_of_bins, double average[], double fit[], float number_of_extrema_profile[], double frc[], double frc_sigma[], int first_fit_bin);
void  RescaleSpectrumAndRotationalAverage(Image* spectrum, Image* number_of_extrema, Image* ctf_values, int number_of_bins, double spatial_frequency[], double average[], double average_fit[], float number_of_extrema_profile[], float ctf_values_profile[], int last_bin_without_aliasing, int last_bin_with_good_fit);
void  Renormalize1DSpectrumForFRC(int number_of_bins, double average[], double fit[], float number_of_extrema_profile[]);
float ReturnAzimuthToUseFor1DPlots(CTF* ctf);
```

#### Suggested refactor:

Although I hate to bloat the Image class any more, that is where these belong. Perhaps the declarations go in src/core/image.h, but the definitions remain in src/programs/ctffind/image_ext/image.cpp

jojoelfe commented:
Hey,

sorry for loosing the ball here a bit. I mostly agree with most points. The question is mainly should we do this in this PR or later?

Regarding moving the pre-processing methods to the image class: Most of these are very specific in that they act on spectra. i.e. a real-space image that contains amplitudes from an FFT. Maybe we could make a `SpectrumImage` class that inherits from `Image` and put the functions there?

jojoelfe commented:
I've removed and moved all the functions insde ctffind_utils into the core library. 

There is a new class `SpectrumImage` that inherits from `Image` and hast these additional functions:

```c++
    float FindRotationalAlignmentBetweenTwoStacksOfImages(Image* other_image, int number_of_images, float search_half_range, float search_step_size, float minimum_radius, float maximum_radius);
    void  GeneratePowerspectrum(CTF ctf_to_apply);
    void  OverlayCTF(CTF* ctf, Image* number_of_extrema, Image* ctf_values, int number_of_bins_in_1d_spectra, double spatial_frequency[], double rotational_average_astig[], float number_of_extrema_profile[], float ctf_values_profile[], Curve* equiphase_average_pre_max, Curve* equiphase_average_post_max);
    void  ComputeRotationalAverageOfPowerSpectrum(CTF* ctf, Image* number_of_extrema, Image* ctf_values, int number_of_bins, double spatial_frequency[], double average[], double average_fit[], double average_renormalized[], float number_of_extrema_profile[], float ctf_values_profile[]);
    void  ComputeEquiPhaseAverageOfPowerSpectrum(CTF* ctf, Curve* epa_pre_max, Curve* epa_post_max);
    void  RescaleSpectrumAndRotationalAverage(Image* number_of_extrema, Image* ctf_values, int number_of_bins, double spatial_frequency[], double average[], double average_fit[], float number_of_extrema_profile[], float ctf_values_profile[], int last_bin_without_aliasing, int last_bin_with_good_fit);
    float DilatePowerspectrumToNewPixelSize(bool resample_if_pixel_too_small, float pixel_size_of_input_image, float target_pixel_size_after_resampling,
                                            int box_size, Image* resampled_power_spectrum, bool do_resampling = true, float stretch_factor = 1.0f);
```

I do like this change as it also makes code more clear by signaling that this will hold not a micrograph, but a powerspectrum.

The functions

```c++
    float ReturnAzimuthToUseFor1DPlots( );
    void  ComputeImagesWithNumberOfExtremaAndCTFValues(Image* number_of_extrema, Image* ctf_values);
```

have been added to the CTF class.

These thrre functions don't really operate on classes and therefore were harder to place. I put them for now into the ctf.h/ctf.cpp file, but outside of the class

```c++
void Renormalize1DSpectrumForFRC(int number_of_bins, double average[], double fit[], float number_of_extrema_profile[]);
void ComputeFRCBetween1DSpectrumAndFit(int number_of_bins, double average[], double fit[], float number_of_extrema_profile[], double frc[], double frc_sigma[], int first_fit_bin);
int  ReturnSpectrumBinNumber(int number_of_bins, float number_of_extrema_profile[], Image* number_of_extrema, long address, Image* ctf_values, float ctf_values_profile[]);
```

@bHimes @timothygrant80  you want to take another look.

The comparison classes and objective functions are very tightly coupled to the ctffind algorithm, so I couldn't find a straightforward way of moving them into the core library.

bHimes reviewed (commented):
So what happens when the input is anywhere between 0.00001 and 0.01? The answer is not defined is it?

Furthermore,
I guess this must be used only in FFT's of real images where the input is always positive, but if the input were a complex image (as in the simulator wave propagation) this would fail for small negative numbers.

Suggesting

```cpp
inline float sinc(float radians) {
    // The value of 0.01 is probably not the best choice - shouldn't it 
    // just be the smallest normal float?
    constexpr float constant_threshold_value = 0.01;
    if ( fabsf(radians) > constant_threshold_value )
        return sinf(radians) / radians;
    else
        return 1.0f;
}
```

bHimes commented on src/core/spectrum_image.cpp:
```diff
@@ -0,0 +1,671 @@
+    float       j_logi, j_logi_sq, i_logi, i_logi_sq;
+    float       current_spatial_frequency_squared;
+    long        address                     = 0;
+    const float inverse_logical_x_dimension = 1.0 / float(logical_x_dimension);
```
Isn't this already an object property (Image::fourier_voxel_size_x(y,z))

bHimes commented on src/core/spectrum_image.cpp:
```diff
@@ -0,0 +1,671 @@
+            // Where are we?
+            current_spatial_frequency_squared = j_logi_sq + i_logi_sq;
+
+            current_azimuth   = atan2f(j_logi, i_logi);
```
Many other places in the code have a logical to check for zero frequency and return zero which I assume is to avoid undefined behavior for atan2f under some implementations?

jojoelfe commented:
> So what happens when the input is anywhere between 0.00001 and 0.01? The answer is not defined is it?

Then the final line takes over, which I guess is a polynomial approximation (I think it comes from the boost library). 
`return 1.0 - temp_float / 6.0 + temp_float * temp_float / 120.0;`

bHimes reviewed (dismissed):
Just a few very minor things jumped out at me. They might not be anything you changed as it is hard to see what is just being moved vs what is new code. Going to trust that you've thought through the rest!

jojoelfe commented:
> Just a few very minor things jumped out at me. They might not be anything you changed as it is hard to see what is just being moved vs what is new code. Going to trust that you've thought through the rest!

Yes, all the function have just been copy-pastes (and the first argument removed and replaced by `this`). 

The exception is that small change to the sinc function, which seems to remove the gcc test failure.

bHimes commented:
> > So what happens when the input is anywhere between 0.00001 and 0.01? The answer is not defined is it?
> 
> Then the final line takes over, which I guess is a polynomial approximation (I think it comes from the boost library). `return 1.0 - temp_float / 6.0 + temp_float * temp_float / 120.0;`

Oops, was only seeing the collapsed view. Odd choise by GH to truncate two lines of a function...
Yes, this comes from the boost library (or whatever source they used)

bHimes commented:
> > Just a few very minor things jumped out at me. They might not be anything you changed as it is hard to see what is just being moved vs what is new code. Going to trust that you've thought through the rest!
> 
> Yes, all the function have just been copy-pastes (and the first argument removed and replaced by `this`).
> 
> The exception is that small change to the sinc function, which seems to remove the gcc test failure.

Perfect, thanks for clarifying. The simplest change would be to take the absolute value of the input angle before any of the other code. Since sinc is an even function, this will return the correct results for negative angles.

bHimes reviewed (approved):
lgtm
